engine/src/main/java/org/apache/apex/engine/plugin/ApexPluginManager.java

```diff
@@ -0,0 +1,195 @@
+import com.datatorrent.stram.plan.logical.LogicalPlan;
+import com.datatorrent.stram.webapp.AppInfo;
+
+/**
+ * A default implementation for ApexPluginDispatcher. It handler common tasks such as per handler
+ * registration. actual dispatching is left for classes extending from it.
+ */
+public abstract class ApexPluginManager extends AbstractService
+{
+  private static final Logger LOG = LoggerFactory.getLogger(ApexPluginManager.class);
+  protected final Collection<ApexPlugin> plugins = Lists.newArrayList();
+  protected final StramAppContext appContext;
+  protected final StreamingContainerManager dmgr;
+  private final PluginLocator locator;
+  private final AppInfo.AppStats stats;
+  protected Configuration launchConfig;
+  protected FileContext fileContext;
+  protected final Map<ApexPlugin, PluginInfo> pluginInfoMap = new HashMap<>();
+  protected PluginContext pluginContext;
+
+  public ApexPluginManager(PluginLocator locator, StramAppContext context, StreamingContainerManager dmgr, AppInfo.AppStats stats)
+  {
+    super(ApexPluginManager.class.getName());
+    this.locator = locator;
+    this.appContext = context;
+    this.dmgr = dmgr;
+    this.stats = stats;
+    LOG.debug("Creating apex service ");
+  }
+
+  private Configuration readLaunchConfiguration() throws IOException
+  {
+    Path appPath = new Path(appContext.getApplicationPath());
+    Path  configFilePath = new Path(appPath, LogicalPlan.LAUNCH_CONFIG_FILE_NAME);
+    try {
+      LOG.debug("Reading launch configuration file ");
+      URI uri = appPath.toUri();
+      Configuration config = new YarnConfiguration();
+      fileContext = uri.getScheme() == null ? FileContext.getFileContext(config) : FileContext.getFileContext(uri, config);
+      FSDataInputStream is = fileContext.open(configFilePath);
+      config.addResource(is);
+      LOG.debug("Read launch configuration");
+      return config;
+    } catch (FileNotFoundException ex) {
+      LOG.warn("Configuration file not found {}", configFilePath);
+      return new Configuration();
+    }
+  }
+
+  @Override
+  protected void serviceInit(Configuration conf) throws Exception
+  {
+    super.serviceInit(conf);
+    this.launchConfig = readLaunchConfiguration();
+    if (locator != null) {
+      Collection<ApexPlugin> plugins = locator.discoverPlugins(conf);
+      if (plugins != null) {
+        this.plugins.addAll(plugins);
+      }
+    }
+
+    for (ApexPlugin plugin : plugins) {
+      plugin.setup(new PluginManagerImpl(plugin));
+    }
+  }
+
+  @Override
+  protected void serviceStop() throws Exception
+  {
+    for (ApexPlugin plugin : plugins) {
+      plugin.teardown();
+    }
+    super.serviceStop();
+  }
+
+  /**
+   * Keeps information about plugin and its registrations. Dispatcher use this
+   * information while delivering events to plugin.
+   */
+  class PluginInfo
+  {
+    private final ApexPlugin plugin;
+    private final Map<RegistrationType<?>, List<?>> registrationMap = new HashMap<>();
```

Comment:
In which scenarios would a single plugin register multiple handlers for the same type?